src/treebuilder.rs

```diff
@@ -0,0 +1,223 @@
+use libc::{c_int, c_void};
+use std::marker;
+use std::ffi::OsStr;
+
+use {panic, raw, Error, Oid, Repository, Tree, TreeEntry};
+use util::{Binding, IntoCString};
+
+/// Constructor for in-memory trees
+pub struct TreeBuilder<'repo> {
+    raw: *mut raw::git_treebuilder,
+    _marker: marker::PhantomData<&'repo Repository>,
+}
+
+impl<'repo> TreeBuilder<'repo> {
+    /// Create a new TreeBuilder with no entries
+    pub fn new(repo: &'repo Repository) -> Result<TreeBuilder, Error> {
+        unsafe {
+            let mut ret = 0 as *mut raw::git_treebuilder;
+            try_call!(raw::git_treebuilder_new(&mut ret, repo.raw(),
+                                               0 as *mut raw::git_tree));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Create a new TreeBuilder initialized with the entries of the given Tree
+    pub fn from_tree(repo: &'repo Repository, tree: &Tree<'repo>
+                     ) -> Result<TreeBuilder<'repo>, Error> {
+        unsafe {
+            let mut ret = 0 as *mut raw::git_treebuilder;
+            try_call!(raw::git_treebuilder_new(&mut ret, repo.raw(),
+                                               tree.raw()));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Clear all the entries in the builder
+    pub fn clear(&mut self) {
+        unsafe { raw::git_treebuilder_clear(self.raw) }
+    }
+
+    /// Get the number of entries
+    pub fn len(&self) -> usize {
+        unsafe { raw::git_treebuilder_entrycount(self.raw) as usize }
+    }
+
+    /// Get en entry from the builder from its filename
+    pub fn get<P: AsRef<OsStr>>(&self, filename: P) -> Result<TreeEntry, Error> {
```

Comment:
Thinking about this some more, `AsRef<[u8]>` is probably more appropriate than `AsRef<OsStr>` here since the query functions all return either `&str` or `&[u8]`.  I'll make that change tomorrow.